### PR TITLE
Performance fixes & more flags

### PR DESCRIPTION
**Pull request by mnunberg**

### CODE REVIEW

File: src/buffer.h [hunk 1]
```diff
@@ -27,13 +27,37 @@ typedef struct {
...
+static inline int Buffer_AtEnd(const Buffer *ctx) {
+  return ctx->offset >= ctx->cap;
+}
+
+/**
+Skip forward N bytes, returning the resulting offset on success or the end
+position if where is outside bounds
+*/
+static inline size_t Buffer_Skip(BufferReader *br, int bytes) {
+  // if overflow - just skip to the end
```
Comment (dvirsky) [inline src/buffer.h]: remove comment - this is not the case anymore...

File: src/inverted_index.c [hunk 2]
```diff
@@ -119,72 +163,123 @@ inline int IR_HasNext(void *ctx) {
...
+      IndexReader_AdvanceBlock(ir);
     }
 
-    readEntry(br, ir->readFlags, ir->record, ir->singleWordMode);
+    int rv = ir->decoder(&ir->br, ir->decoderCtx, ir->record);
     ir->lastId = ir->record->docId += ir->lastId;
-
-    // The record doesn't match the field filter. Continue to the next one
-    if (!(ir->record->fieldMask & ir->fieldMask)) {
+    if (!rv) {
```
Comment (dvirsky) [inline src/inverted_index.c]: comment on what's going on here and what is rv

File: src/inverted_index.h [hunk 3]
```diff
@@ -28,9 +28,17 @@ typedef struct {
   uint32_t numDocs;
 } InvertedIndex;
 
+struct indexReadCtx;
+
+typedef union {
```
Comment (dvirsky) [inline src/inverted_index.h]: document what this thingie is and what it's used for - after all it's an API used from the outside...

File: src/inverted_index.h [hunk 4]
```diff
@@ -28,9 +28,17 @@ typedef struct {
...
+  void *ptr;
+  unsigned long long num;
```
Comment (dvirsky) [inline src/inverted_index.h]: uint32_t

File: src/inverted_index.h [hunk 5]
```diff
@@ -28,9 +28,17 @@ typedef struct {
...
+} IndexDecoderCtx;
+
 InvertedIndex *NewInvertedIndex(IndexFlags flags, int initBlock);
 void InvertedIndex_Free(void *idx);
 int InvertedIndex_Repair(InvertedIndex *idx, DocTable *dt, uint32_t startBlock, int num);
+typedef int (*IndexDecoder)(BufferReader *br, IndexDecoderCtx ctx, RSIndexResult *res);
```
Comment (dvirsky) [inline src/inverted_index.h]: documentation note on decoders

File: src/inverted_index.h [hunk 6]
```diff
@@ -82,6 +92,8 @@ int IR_GenericRead(IndexReader *ir, RSIndexResult *res);
 /* Read an entry from an inverted index into RSIndexResult */
 int IR_Read(void *ctx, RSIndexResult **e);
 
+void IR_ReadEx(void *ctx, int (*cb)(const RSIndexResult *res, void *arg), void *arg);
```
Comment (dvirsky) [inline src/inverted_index.h]: remove

File: src/tests/Makefile [hunk 7]
```diff
@@ -39,6 +37,8 @@ ext_example:
 	$(MAKE) -C ./ext-example clean all
 .PHONY: ext_example
 
+bench: bench.o
```
Comment (dvirsky) [inline src/tests/Makefile]: call it benchmark_decoder

File: src/buffer.h (hunk 1 above)
Comment (mnunberg) [inline src/buffer.h]: done

File: src/inverted_index.c (hunk 2 above)
Comment (mnunberg) [inline src/inverted_index.c]: done

File: src/inverted_index.h (hunk 4 above)
Comment (mnunberg) [inline src/inverted_index.h]: done

File: src/inverted_index.h (hunk 5 above)
Comment (mnunberg) [inline src/inverted_index.h]: Done

File: src/inverted_index.h (hunk 6 above)
Comment (mnunberg) [inline src/inverted_index.h]: Done

File: src/tests/Makefile (hunk 7 above)
Comment (mnunberg) [inline src/tests/Makefile]: Done

File: src/inverted_index.h (hunk 3 above)
Comment (mnunberg) [inline src/inverted_index.h]: Done

**Pull request merged**